src/NuGet.Core/NuGet.Protocol.Core.v2/DownloadResourceV2.cs

```diff
@@ -38,34 +40,58 @@ public DownloadResourceV2(V2Resource resource)
                 throw new ArgumentNullException(nameof(identity));
             }
 
+            string uri = string.Empty;
+
             return Task.Run(() =>
             {
                 token.ThrowIfCancellationRequested();
 
-                try
+                for (int i = 0; i < 3; i++)
                 {
-                    var sourcePackage = identity as SourcePackageDependencyInfo;
-                    var repository = V2Client as DataServicePackageRepository;
+                    try
+                    {
+                        var sourcePackage = identity as SourcePackageDependencyInfo;
+                        var repository = V2Client as DataServicePackageRepository;
+
+                        bool isFromUri = repository != null
+                            && sourcePackage?.PackageHash != null
+                            && sourcePackage?.DownloadUri != null;
+
+                        if (isFromUri)
+                        {
+                            uri = sourcePackage?.DownloadUri.AbsoluteUri;
+                        }
+                        else
+                        {
+                            uri = (repository as DataServicePackageRepository)?.Source ?? "Folder";
+                        }
 
-                    if (repository != null
-                        && sourcePackage?.PackageHash != null
-                        && sourcePackage?.DownloadUri != null)
+                        if (isFromUri)
+                        {
+                            // If this is a SourcePackageDependencyInfo object with everything populated
+                            // and it is from an online source, use the machine cache and download it using the
+                            // given url.
+                            return DownloadFromUrl(sourcePackage, repository, token);
+                        }
+                        else
+                        {
+                            // Look up the package from the id and version and download it.
+                            return DownloadFromIdentity(identity, V2Client, token);
+                        }
+                    }
+                    catch (IOException ex) when (ex.InnerException is SocketException && i < 2)
                     {
-                        // If this is a SourcePackageDependencyInfo object with everything populated
-                        // and it is from an online source, use the machine cache and download it using the
-                        // given url.
-                        return DownloadFromUrl(sourcePackage, repository, token);
+                        string message = $"Error downloading {identity} from {uri} {ExceptionUtilities.DisplayMessage(ex)}";
+
+                        Logger.Instance.LogWarning(message);
```

Comment:
This will end up going to the null logger in VS I believe, which is okay I would say.